### PR TITLE
Midterm CinemaApp pull request

### PR DESCRIPTION
**Pull request by looneydevs**
**New features:**

- User authentication forms
- Sessions data rendering
- Rendering and handling seats in the cinema

**What should be done:**

- Render authentication form errors
- ~~Add individual seats' categories styles~~
- Add mobile friendly cinema seats markup
- Add time to choose on Session page
- ~~Add authentication check on Session page~~
- Redirect from main page to movie page with search form data
- ~~Add dynamic session info: total price, number of available seats~~, reservation timer
- ~~Show modal window on tickets purchasing~~
- ~~Add additional services and order submission functionality~~
- ~~Add UserProfile page basic markup: user data, ordered tickets, upcoming and ended events~~

### CODE REVIEW

File: src/components/MovieItem.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+const MovieItem = ({ data }: { data: any }) => {
+  return (
+    <Container>
+      <StyledTitle>{data.original_title}</StyledTitle>
+      <StyledPoster
+        src={`https://image.tmdb.org/t/p/w500${data.poster_path}`}
+        alt={data.original_title}
+      />
+      <DefaultButton text="Details" to={`/movies/${data.id.toString()}`} />
```
Comment (SobolevskyDmitry) [inline src/components/MovieItem.tsx]: :question:

File: src/components/MovieItem.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+const StyledPoster = styled.img`
+  width: auto;
+  height: auto;
+  max-width: 200px;
+  align-self: center;
+`;
+
+const MovieItem = ({ data }: { data: any }) => {
+  return (
```
Comment (SobolevskyDmitry) [inline src/components/MovieItem.tsx]: can be simplified

File: src/components/PopUpSnackbar.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+          >
+            <CloseIcon />
+          </IconButton>
+        ]}
+        {...other}
+      />
+    </Snackbar>
+  );
+
+  const domNode: any = document.getElementById('snackbars');
```
Comment (SobolevskyDmitry) [inline src/components/PopUpSnackbar.tsx:95]: :smiley:

File: src/components/SessionsTable.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  width: 100%;
+`;
+
+const WideCell = styled(TableCell)`
+  grid-column: 1 / 5;
+`;
+
+const SessionsTable = ({ sessions, isAuth }: any) => {
+  const getRows = () => {
+    if (sessions.length === 0) {
```
Comment (SobolevskyDmitry) [inline src/components/SessionsTable.tsx]: can be simplified

File: src/components/bonus/BonusContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  pickedBonuses,
+  loadedBonuses,
+  handleBonusesUpdate
+}: {
+  pickedBonuses: any;
+  loadedBonuses: any;
+  handleBonusesUpdate: any;
+}) => {
+  const renderBonuses = () => {
+    return Object.keys(loadedBonuses).map((key: string) => (
```
Comment (SobolevskyDmitry) [inline src/components/bonus/BonusContainer.tsx]: can be simplified

File: src/components/bonus/BonusSelectItem.tsx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          bgColor={addButtonColor}
+          icon={<AddIcon />}
+          type="add"
+          handleClick={handleBonusClick}
+        />
+        <RoundButton
+          bgColor={removeButtonColor}
+          icon={<RemoveIcon />}
+          type="remove"
+          disabled={value === 0}
```
Comment (SobolevskyDmitry) [inline src/components/bonus/BonusSelectItem.tsx]: can be simplified

File: src/components/buttons/HeaderButton.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+interface Props {
+  text: string;
+  to?: string;
+  handleClick?: any;
+  icon?: any;
+  disabled?: boolean;
+}
+
+const Container = styled.div<any>``;
```
Comment (SobolevskyDmitry) [inline src/components/buttons/HeaderButton.tsx]: :smile:

File: src/components/fields/FieldContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      return null;
+  }
+};
+
+const FieldContainer = (props: Props) => (
+  <Grid container={true} spacing={8} justify="center" alignItems="center">
+    <Grid item={true}>
+      <IconStyled icon={props.icon || props.entity} />
+    </Grid>
+    <Grid item={true} lg={7}>
```
Comment (SobolevskyDmitry) [inline src/components/fields/FieldContainer.tsx]: move to local const

File: src/components/seats/SeatsScheme.tsx [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+      console.log(error);
+    }
+  };
+
+  useEffect(() => {
+    loadSeats();
+  }, [hallId]);
+
+  const renderSeats = () => {
+    if (!seats || seats.length === 0) return 'No seats found';
```
Comment (SobolevskyDmitry) [inline src/components/seats/SeatsScheme.tsx]: seats.length === 0 :cry:

File: src/components/seats/SeatsMenu.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  }
+`;
+
+const SeatsMenu = ({
+  onHallChange,
+  onOptionsChange,
+  options,
+  hallSelected
+}: any) => {
+  const handleCheckboxChange = (e: any) => {
```
Comment (SobolevskyDmitry) [inline src/components/seats/SeatsMenu.tsx]: can be simplified

File: src/components/seats/OrderController.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+  const handleSnackbar = (message: string, variant: string) => {
+    setSnackbarInfo({
+      isOpen: true,
+      variant,
+      message
+    });
+  };
+
+  const isEmpty = order.seatsPicked.length === 0;
```
Comment (SobolevskyDmitry) [inline src/components/seats/OrderController.tsx]: can be simplified

File: src/components/seats/SeatsContainer.tsx [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+      }
+    });
+    setOptions(newOptions);
+  };
+
+  const handleSeatPick = (e: any) => {
+    const { seatsPicked } = order;
+    const pickedRow = +e.target.dataset.row;
+    const pickedSeat = +e.target.dataset.seat;
+    const free = e.target.dataset.free === 'true';
```
Comment (SobolevskyDmitry) [inline src/components/seats/SeatsContainer.tsx:101]: :+1:

File: src/components/seats/SeatsContainer.tsx [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    const pickedBefore = seatsPicked.some(
+      (seat: any) => seat.row === pickedRow && seat.seat === pickedSeat
+    );
+    if (!pickedBefore) {
+      newSeatsPicked = seatsPicked.concat({
+        row: pickedRow,
+        seat: pickedSeat,
+        price
+      });
+    } else {
```
Comment (SobolevskyDmitry) [inline src/components/seats/SeatsContainer.tsx]: ternary operator

File: src/components/seats/elements/SeatItem.tsx [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+  row,
+  seat,
+  price,
+  category,
+  isSelected,
+  isReserved,
+  isOrdered,
+  isMuted
+}: any) => {
+  const computeColors = () => {
```
Comment (SobolevskyDmitry) [inline src/components/seats/elements/SeatItem.tsx]: https://www.npmjs.com/package/classnames

File: src/pages/Home.tsx [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  const handleSubmit = (e: any) => {
+    e.preventDefault();
+    console.log(movie);
+    console.log(city);
+    console.log(cinema);
+    console.log(date);
+  };
+
+  useEffect(() => {
+    if (movie !== '' || city !== '' || cinema !== '') {
```
Comment (SobolevskyDmitry) [inline src/pages/Home.tsx]: :cry:

File: src/pages/Movies.tsx [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+const MoviesContainer = styled.div`
+  margin: 0 auto;
+  max-width: 80%;
+  display: flex;
+  flex-wrap: wrap;
+  justify-content: center;
+`;
+
+class Movies extends React.Component<any, any> {
```
Comment (SobolevskyDmitry) [inline src/pages/Movies.tsx]: lets use hooks here, too

File: src/pages/Movies.tsx [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    this.setState({
+      filterText: e.target.value.toLowerCase().trim()
+    });
+  };
+
+  getMoviesList = () => {
+    const { movies } = this.props;
+    const { filterText } = this.state;
+    const filteredMovies =
+      filterText === ''
```
Comment (SobolevskyDmitry) [inline src/pages/Movies.tsx]: :cry:

File: src/pages/Movies.tsx [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        ? movies
+        : movies.filter((movie: any) =>
+            movie.original_title.toLowerCase().includes(filterText)
+          );
+    if (filteredMovies.length === 0) {
```
Comment (SobolevskyDmitry) [inline src/pages/Movies.tsx]: :sob:

File: src/components/seats/SeatsContainer.tsx (hunk 12 above)
Comment (looneydevs) [reply src/components/seats/SeatsContainer.tsx:101]: `!!'false' === true // true`

**Pull request merged**